### PR TITLE
feat: add zilliqa schnorr signing

### PR DESCRIPTION
**Pull request by feri42**
Add schnorr for Zilliqa. To be used in seed signer: https://github.com/ExodusMovement/exodus-hydra/pull/10588

Toward: https://github.com/ExodusMovement/assets/issues/2092

### CODE REVIEW

File: features/keychain/module/crypto/schnorr-z.js
Comment (feri42) [reply features/keychain/module/crypto/schnorr-z.js:4]: Yes. We can directly import `@exodus/keychain/module/crypto/secp256k1` for use in the library.

Comment (feri42) [reply features/keychain/module/crypto/schnorr-z.js:4]: Unless you have a different suggestion?

Comment (fboucquez) [reply features/keychain/module/crypto/schnorr-z.js:4]: Maybe in the future, an asset won't be able/won't need to define signing, signing is always happing via the keychain.  (move funds/import key would break this logic though)

Comment (feri42) [reply features/keychain/module/crypto/schnorr-z.js:4]: We would need to add support for adding a private key to the keychain. Not sure what that would look like. Currently the keychain operates by deriving private keys from a seed. This private key would not be associated with a seed.

**Review by fboucquez — DISMISSED** [review]

**Comment by mvayngrib** [comment]
i feel unqualified to review this 😅 , @ChALkeR @kewde ?

**Comment by feri42** [comment]
> i feel unqualified to review this 😅 , @ChALkeR @kewde ?

The code is really just copied from here: https://github.com/ExodusMovement/assets/blob/a8fec3450079e64d5fdff5600fa5f4f0aca0d111/zilliqa/zilliqa-lib/src/schnorr.js#L37 . Don't think we need to review the cryptography itself. But I do need at least concept approval.

**Review by mvayngrib — DISMISSED** [review]
utACK

compared with https://github.com/ExodusMovement/assets/blob/a8fec3450079e64d5fdff5600fa5f4f0aca0d111/zilliqa/zilliqa-lib/src/schnorr.js

**Comment by mvayngrib** [comment]
how are zilliqa schnorr sigs diff, can we just use regular schnorr?

**Comment by feri42** [comment]
> how are zilliqa schnorr sigs diff, can we just use regular schnorr?

I won't pretend to completely understand the differences, but the way I see it is the normal schnorr gets the random entropy once through the `extraEntropy` parameter and uses a hash of that random value while internally iterating.

The zilliqa schnorr computes a new random value on every iteration.

I'm not even sure there is a quantitative difference there.

**Review by ChALkeR — COMMENTED** [review]
Concept ack (discussed a while ago)

Did not closely review the code (yet)

Not sure if we should place this in /crypto, but it doesn't block this PR

Nit: given the async nature of keychain, this should use async hmac/sha256 instead, that will make those use native impls in webcrypto contexts

File: features/keychain/module/crypto/schnorr-z.js
Comment (ChALkeR) [inline features/keychain/module/crypto/schnorr-z.js:2]:
```suggestion
import { hash } from '@exodus/crypto/hash'
import { hmac } from '@exodus/crypto/hmac'
```

And turn everything into async

**Comment by ChALkeR** [comment]
> how are zilliqa schnorr sigs diff, can we just use regular schnorr?

@mvayngrib, no, zil has a different schnorr impl
Having a separate impl for it is reasonable (in concept), technical details need to be worked out

File: features/keychain/module/crypto/schnorr-z.js
Comment (feri42) [reply features/keychain/module/crypto/schnorr-z.js]: https://github.com/ExodusMovement/exodus-oss/pull/185/commits/a72f683cd02ebcd6a0b265b61955cd2a4c67b672 .

**Comment by feri42** [comment]
Given that we have Concept ACK, can we have this reviewed?

**Review by mvayngrib — APPROVED** [review]

**Review by fboucquez — APPROVED** [review]

**Pull request merged**
